### PR TITLE
Rails 5.2

### LINKED ISSUE #497: [Gemfile] Update to (at least) Rails 5.1.6
(Note for Hacktoberfest participants: Upgrading Rails is usually a complex, multi-stage process. But it is important to keep a web app up-to-date with recent software that has received security updates and other improvements. Here, the process of updating Rails is laid out into all of the individual stages, so anyone can try to complete the upgrade. I am not expecting this to be easy, but hopefully it will be somewhat educational to go through this process. Don't worry about it if you get stuck. We can help you out if you have any questions. Or feel free to skip this and look at our other issues!)

### Updating the Gemfile

As a warming-up exercise, update [`Gemfile`](https://github.com/RefugeRestrooms/refugerestrooms/blob/develop/Gemfile) so that the line for the `rails` gem ("`gem 'rails', '5.1.4'`") is updated to say `5.1.6` instead of `5.1.4`.

This is the basic first step. To really complete the upgrade, we will have to start over, but it's impossible to complete this upgrade without being comfortable updating the Gemfile.

### Obtaining an updated `Gemfile.lock` using a Docker container

Start over with a `Gemfile` where Rails has not been updated.

Set up a Docker container running the Refuge Restrooms web app, by following along with [`CONTRIBUTING.md`](https://github.com/RefugeRestrooms/refugerestrooms/blob/develop/CONTRIBUTING.md) 

Then, start an interactive terminal/shell session on the Docker container, by running `docker-compose run web bash`.

Next, update `Gemfile` as described above in the "Getting started" section.

Then, in the interactive terminal/shell session, install the updated Rails version (and update `Gemfile.lock`) by running this command: `bundle update rails`.

### Solving dependency conflicts in `Gemfile` / `Gemfile.lock`

When you run `bundle update rails`, there may be error messages about conflicts between different required versions of the same gems. To solve these, you will need to update `Gemfile` so that it has a set of gems where none of the version, or their requirements, are in conflict. The specific error messages will tell you what gems, including dependencies, are in conflict; The error messages will also tell you what range of version numbers are acceptable for those gems. 

(Bonus tip: Instead of updating the `Gemfile`, sometimes you can list all gems that need updating in a `bundle update` command, like so: `bundle update rails grape active-admin [more gems here]`. You may still need to edit the `Gemfile` in a place or two, but this way can result in a much smaller set of changes to the `Gemfile`.)

To find the latest version numbers for all gems in the `Gemfile`, you can run `bundle outdated`. (See this [documentation](https://bundler.io/man/bundle-outdated.1.html) for more info on the `bundle outdated` command.)

As an alternative to running `bundle outdated`, you can look at all past version numbers of a given gem by visiting its page at rubygems.org. For example: https://rubygems.org/gems/rails

### Other issues when updating rails

Sometimes there are updated or deprecated features in a new release of Rails. We are only updating to a "patch" level update in this case, so there shouldn't be any issues besides conflicts in the `Gemfile`. However, there might be other changes needed in our app, such as changes to our configuration files, or updates to the code we use. If you need help with any of these things, maintainers are here and glad to help. Just let us know with a comment on this issue, on your Pull Request, or on [Slack](https://github.com/RefugeRestrooms/refugerestrooms#slack).

## Total completion (testing):

To truly finish this upgrade, we need to test that everything still works after the update.

With the Docker container up and running, run this command to execute our automated testing process: `docker-compose run -e "RAILS_ENV=test" web rake db:test:prepare spec`.

If there are errors, we usually have to go back and edit some configuration files, some code, or the `Gemfile` again.

### Manual testing and finishing up

Sometimes automated testing doesn't give a full enough picture of what is working and what isn't. Ideally, when making a large update like moving to a new version of rails, we would also test manually. Here is how to test the site manually:

With the Docker container running our web app, visit `127.0.0.1:3000` in your web browser. You should see our home page.

(If you are using Docker Toolbox, for older Windows or Mac computers, try visiting `192.168.99.100:3000` instead.)

Then you can visit some pages (or ideally all of them) and click on the main functions of those pages to see if they work. You can submit a new restroom or rate restroom entries within the Docker container without issues, as this is totally separate from our live, "production" restrooms database at Refugerestrooms.org.

### PR DESCRIPTION
**Pull request by btyy77c**
# Context
- Fixes #497
- Update to Rails 5.2

# Summary of Changes

- Put Gems in alphabetical order
- Updated Gems: activeadmin, grape, grape-swagger, grape-kaminari, rails, rspec-rails, simple_form
- Added Gems: bootsnap, listen
- Updated test: changed `expect(response).to be_success` to `assert_response :success`
- Updated config files based on rails app:update recommendations

# Checklist

- [ ] Tested Mobile Responsiveness
- [ ] Added Unit Tests
- [ ] CI Passes
- [ ] Deploys to Heroku on test Correctly (Maintainers will handle)
- [ ] Added Documentation (Service and Code when required)

# Screenshots

## Before
Command-line interface booting Rails. Shows Rails 5.1.4 and Puma 3.11.2

## After

Command-line interface booting Rails. Shows Rails 5.2.1 and Puma 3.12.0

### CODE REVIEW

**Comment by DeeDeeG** [comment]
I notice some of these commits are made with a different git "name" and email set than your official GitHub name/email combo. So not all of these commits would get attributed to your GitHub account properly when we merge.

(This wouldn't make a difference if we squash and merge, since it would produce just the one commit, hopefully with your correct info on it. We tend to decide on whether to squash or not at merge time.)

If you want to go back and fix this (or have us fix it), it should be doable.

Here's two methods: 
https://www.git-tower.com/learn/git/faq/change-author-name-email#chapter_using+interactive+rebase
https://www.git-tower.com/learn/git/faq/change-author-name-email#chapter_using+git+filter-branch

Trying this out in Docker as soon as I get Docker installed on my new Ubuntu install.

**Comment by btyy77c** [comment]
> * The API page isn't working. (This seems to be a canary in the coal mine; It breaks a lot.)

When I visit production, the api is not working: (https://www.refugerestrooms.org/api or https://www.refugerestrooms.org/api.json)   Is there a different route or something else I should be checking? Thanks!

**Comment by DeeDeeG** [comment]
The correct URL is in the Resources dropdown of the top navigation...

Appears to be `[base-url]/api/docs/`.

**Comment by DeeDeeG** [comment]
I think there is something off with our Docker setup, because none of the dummy entries from `db/export.csv` are loaded, even on the `develop` branch.

On `develop`: The API page does work. And if I submit a new restroom, then ask for all restrooms by date, the one I submitted shows up in response. 

[Screenshot of the API page, showing just one result](https://user-images.githubusercontent.com/20157115/46573800-2fc4ab80-c968-11e8-9213-524830748219.png)

**Comment by btyy77c** [comment]
> The API page can be visited at `[base-url]/api/docs/`, e.g.: `127.0.0.1:3000/api/docs`.

So, I'm not sure how long you've worked on this application, but do you know the history of why the `/api/docs` app was placed in the `public folder` (https://github.com/RefugeRestrooms/refugerestrooms/tree/develop/public/api/docs)? Is there a reason it couldn't be part of the rails application?

**Comment by btyy77c** [comment]
Sorry! I didn't mean to close the pull request. I normally use gitlab and I'm apparently terrible with GitHub.

**Comment by DeeDeeG** [comment]
> So, I'm not sure how long you've worked on this application, but do you know the history of why the /api/docs app was placed in the public folder . . .[link]. . . ?

This was before I came in. But I guess it's sort of a hack/kludge that lets this page be viewable without doing a lot of route configuration and so on.

> Is there a reason it couldn't be part of the rails application?

No, I think that is just how the contributor who added this ended up doing it. Ideally, in my opinion, we would go in and install this as a proper nodejs app via our `.package.json`. But "part of the app" sounds like the right direction to go in :+1: .

>Sorry! I didn't mean to close the pull request. I normally use gitlab and I'm apparently terrible with GitHub.

No worries! (It's nice to hear not everyone is on closed platforms, and totally open-source workflows are viable.)

**Comment by DeeDeeG** [comment]
From https://github.com/RefugeRestrooms/refugerestrooms/issues/492#issuecomment-427581913:

> As @tkwidmer mentioned in Slack, we might want to also consider installing swagger-ui as a proper node package.
> 
> Our `public/api` directory is mostly (entirely?) a copy of the `swagger-ui` project from the 2.0 series: https://github.com/swagger-api/swagger-ui/tree/2.x/dist
> 
> I suspect we could install (the 2.0 version?) of the nodejs package and then update as needed.

**Comment by DeeDeeG** [comment]
Sorry to bother you @btyy77c. (These pull requests you have put in are really helpful!)

To clarify about the commit author deatils: Does it bother you if these commits aren't linked to your GitHub account?

If so: Please let us know what e-mail should be in these commits to properly link to your account. I'm prepared to rebase the pull requests with the correct author info if that works for you. You can check what e-mails are already linked to your GitHub account (or even add another email address to your account) here: https://github.com/settings/emails

If not, or if you'd actually prefer these commits not be linked to your account (better anonymity, etc), please let us know. That is no problem, but I'd hate to use someone's work without crediting them, or at least confirming how they do or don't want to be attributed.

Thanks again!

\- DeeDeeG

**Comment by btyy77c** [comment]
> Sorry to bother you @btyy77c. (These pull requests you have put in are really helpful!)
> 
> To clarify about the commit author deatils: Does it bother you if these commits aren't linked to your GitHub account?
> 
> If so: Please let us know what e-mail should be in these commits to properly link to your account. I'm prepared to rebase the pull requests with the correct author info if that works for you. You can check what e-mails are already linked to your GitHub account (or even add another email address to your account) here: https://github.com/settings/emails
> 
> Edit: As a reminder, there is a totally anonymous email option that still links these commits to your GitHub account. That would be the `username@users.noreply.github.com` format.
> 
> If not, or if you'd actually prefer these commits not be linked to your account (better anonymity, etc), please let us know. That is no problem, but I'd hate to use someone's work without crediting them, or at least confirming how they do or don't want to be attributed.
> 
> Thanks again!
> 
> - DeeDeeG

I don't have a preference either way.  I use gitlab mostly and I don't want to change my email to anything like `noreply.github.com`. Thanks for checking first!

**Comment by DeeDeeG** [comment]
Hi again,

I did get to take another look at this. The primary issue left with this pull request is the API documentation page not working.

I am pretty confident that has to do with our API being built against an older API spec, and the new `grape` and `grape-swagger` gems not being compatible with our somewhat old API.

We would need to update our actual API code to fix this; I made issue #516 to address it.

For now I'd recommend rolling back `grape` and `grape-swagger` to the versions they have in the current `develop` branch. That should make the API page functional again.

I'm also taking a look at the way the new config files are merged, but everything generally works as expected when I boot this up in Docker, so the config files are not likely to need changes, or else they would just be small tweaks.

Best,

\- DeeDeeG

**Comment by DeeDeeG** [comment]
Well, the updated `grape` and `grape-swagger` gems are working in #506, so clearly our API is fundamentally compatible with the new gem versions.

I will try to see if there is any way to get the new swagger setup working on Rails 5.2.1, and if so I will update here. Then we can probably merge this and #506 once we get them working together.

**Comment by DeeDeeG** [comment]
For some reason, on this branch I get two requests for "/api/swagger_doc.json" every time I visit the API docs page:

```logtalk
Started GET "/api/swagger_doc.json" for 10.0.2.2 at 2018-10-16 02:42:03 +0000
Started GET "/api/swagger_doc.json" for 10.0.2.2 at 2018-10-16 02:42:03 +0000
```

**Comment by DeeDeeG** [comment]
In the process of troubleshooting this pull request, I went and performed a Rails 5.2.1 upgrade over again from scratch in this branch: https://github.com/DeeDeeG/refugerestrooms/tree/rails-5.2.1

It's a bit more minimal in terms of changes, and it does work, including the API documentation page.

I only updated the minimum amount of Gems required, didn't enable the `listen` Gem, and tried to keep config changes to a minimum and preserve existing custom configs, while merging the new ones from 5.2.1.

I still can't figure out why this pull request broke the API page. After all of this, I might lean toward just merging [that branch](https://github.com/DeeDeeG/refugerestrooms/tree/rails-5.2.1). But the proof that this was easy enough to do in an afternoon, and some impetus to actually get this done, is still owed to this pull request, regardless. So thanks again.

Best regards,

\- DeeDeeG

**Comment by DeeDeeG** [comment]
I (finally) figured out what was going on with the API page.

This branch still has the old `2.x` version of `swagger-ui`, which only supports APIs made conforming to the  Swagger 1.2 API specification...

Upgrading the `grape` and `grape-swagger` gems causes our app to serve an OpenAPI 2.0 API... which the old swagger-api code doesn't recognize and assumes is "deprecated", and ultimately failing to load. 

So that match-up is what breaks the docs page. (Evidently the API itself did not break, just the page explaining the API to end-users. That's why our tests all passed anyway. I think we should add some RSpec tests to the API docs page at some point, so we test that page, and not just the raw API calls.)

---

In short: We can fix this pull request by using older `grape` and `grape-swagger`. (See [this branch at my fork](https://github.com/DeeDeeG/refugerestrooms/commits/rails52-with-older-grape-grape-swagger).)
